### ISSUE TITLE
flash_attn does not apply mask in SpatialFeatureAttention module

### ISSUE THREAD

**Issue by nitish-nayak** [NVlabs/WarpConvNet#26]
We noticed that the [SpatialFeatureAttention](https://github.com/NVlabs/WarpConvNet/blob/main/warpconvnet/nn/modules/attention.py#L142) module applies `flash_attn_qkvpacked_func` on the full padded tensor without any prior masking as shown below. This goes before the softmax and hence the attention on the active tokens includes effects from the padded tokens. 
```
            # Flash attention path
            # Flash attention - preserve original dtype if possible
            original_dtype = qkv.dtype
            if qkv.dtype not in [torch.float16, torch.bfloat16]:
                # Convert to half precision for flash attention
                qkv_flash = qkv.half()
            else:
                qkv_flash = qkv

            x = flash_attn.flash_attn_qkvpacked_func(
                qkv_flash,
                dropout_p=self.attn_drop_p if self.training else 0.0,
                softmax_scale=self.scale,
            ).reshape(B, N, C)

            # Convert back to original dtype if necessary
            if x.dtype != original_dtype:
                x = x.to(original_dtype)

        x = self.proj(x)
        x = self.proj_drop(x)
```
In addition, the non-flash path of the same module just seems to add the mask tensor directly, which just adds `1` to the active tokens and `0` to the padded tokens. Therefore, this too includes effects from the padded tokens on the active token attention score. 
```
            if mask is not None:
                attn = attn + mask

            attn = attn.softmax(dim=-1)
```

Is there a reason why its done this way here? In the `PatchAttention` module, it seems like its done more correctly without going to the padded representation by using`flash_attn_varlen_qkvpacked_func` on the concatenated tensor directly. 
```
        out_feat = flash_attn.flash_attn_varlen_qkvpacked_func(
            qkv,
            attn_offsets,
            max_seqlen=K,
            dropout_p=self.attn_drop_p if self.training else 0.0,
            softmax_scale=self.scale,
        )
```
Shouldn't this approach be also used for `SpatialFeatureAttention`? 

Thank you for the great project and your time!

**Comment by chrischoy** [comment]
Thanks for the issue. Fixed https://github.com/NVlabs/WarpConvNet/commit/66b3cf960177e142111668bd411d276c988fc6be
